### PR TITLE
[Serializer] Add local cache to normalizers

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #24206
| License       | MIT
| Doc PR        | -

Should help making the Serializer a bit faster.

### CODE REVIEW

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php [hunk 1]
```diff
@@ -35,13 +35,14 @@
...
     private static $setterAccessibleCache = array();
+    private $cache = array();
 
     /**
      * {@inheritdoc}
      */
     public function supportsNormalization($data, $format = null)
     {
-        return parent::supportsNormalization($data, $format) && $this->supports(get_class($data));
+        return parent::supportsNormalization($data, $format) && (isset($this->cache[$type = \get_class($data)]) ? $this->cache[$type] : $this->cache[$type] = $this->supports($type));
```
Comment (ro0NL) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: applies to supportsDenormalization as well, no?

Comment (ro0NL) [reply src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: what about `private static $cache` + `private static supports()` + inline isGetMethod

File: src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php [hunk 2]
```diff
@@ -30,12 +30,14 @@
...
 {
+    private $cache = array();
+
     /**
      * {@inheritdoc}
      */
     public function supportsNormalization($data, $format = null)
     {
-        return parent::supportsNormalization($data, $format) && $this->supports(get_class($data));
+        return parent::supportsNormalization($data, $format) && (isset($this->cache[$type = \get_class($data)]) ? $this->cache[$type] : $this->cache[$type] = $this->supports($type));
```
Comment (ro0NL) [inline src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php]: same

**Review by ro0NL — COMMENTED** [review]
\+ inline `Serializer::denormalizeObject()`?

**Comment by nicolas-grekas** [comment]
> + inline Serializer::denormalizeObject()

Done.
For other comment: static state won't provide anything is practice, and are not better in terms of design.

File: src/Symfony/Component/Serializer/Normalizer/ArrayDenormalizer.php [hunk 3]
```diff
@@ -66,7 +66,7 @@ public function denormalize($data, $class, $format = null, array $context = arra
      */
     public function supportsDenormalization($data, $type, $format = null/*, array $context = array()*/)
     {
-        $context = func_num_args() > 3 ? func_get_arg(3) : array();
+        $context = \func_num_args() > 3 ? func_get_arg(3) : array();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/ArrayDenormalizer.php]: see https://github.com/php/php-src/blob/f2db305fa4e9bd7d04d567822687ec714aedcdb5/Zend/zend_compile.c#L3872

**Review by ro0NL — APPROVED** [review]

**Review by ogizanagi — APPROVED** [review]

**Comment by ogizanagi** [comment]
Won't it make sense to backport what could be on lower branches?

**Comment by nicolas-grekas** [comment]
> Won't it make sense to backport what could be on lower branches?

AFAIK, we merge perf improvements as new feats usually (even if we diverged sometime from this rule, but we shouldn't anymore unless catastrophic. Not sure it's the case here?)

**Pull request merged**